cisagov/getgov
# Save application

## 🗣 Description ##

This PR adds all the plumbing needed to save and restore progress on domain applications.

## 💭 Motivation and context ##

This is a very big PR with a number of outstanding items and caveats.

- [ ] Add in progress and submitted examples of applications to our data fixtures
    - [ ] On that note, it would be nice to have our fixtures autoload whenever the docker containers start up
- [ ] Complete the unit tests for restore; they are currently marked with `@skip`
- [ ] A messaging system needs to be created and tested; currently the user has no way to know that pressing "save" did anything at all
- [ ]  There is an edge case where a user can select "Federal", click next, select an option ("Judicial", for example), click save or next, then go back and select, say, "State", and click next. Their database record will now say that they are of type "State" and of federal type "Judicial". This seems like a good candidate for `wontfix`.
- [ ] Other contacts, organization website, and alternative domains currently have no mechanism to add more than one
- [ ] Lastly, the combo boxes don't work as expected; the correct value is being "selected" in the underlying `<select>` element but does not show up visually on the page. I am in favor of switching to either a simple select or text input due to [other accessibility issues](https://github.com/uswds/uswds-site/issues/1898) in this component, but I'd like to get your opinion, @igorkorenfeld
- [ ]  Have `Domain` inherit from `Website` so that they can share all code related to the domain name string

Closes #260

## 🧪 Testing Instructions ##

Manual testing of save is quite easy:
1. Go to http://localhost:8080/register to begin a new application
2. Open the admin panel  http://localhost:8080/admin/registrar/domainapplication/ in another tab
3. On each page, fill in the fields and click save
4. Verify that the page refreshes and that the info you saved is now in the database model

Manual testing of restore is a bit fiddly. You'll need to create an application, fill in most fields, save it (but don't finish it!), then delete your cookies, then visit http://localhost:8080/application/1/edit (assuming the domain application primary key is `1`).
